sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/ServerCertificateCheck.kt

```diff
@@ -19,44 +19,107 @@
 
-        val extendsX509 = bindingContext.get(CLASS, node)?.getAllSuperClassifiers()?.any {
-            it.fqNameOrNull()?.asString() == "javax.net.ssl.X509TrustManager"
-        } ?: return
-        if (extendsX509) {
-            node.body?.functions?.forEach { f ->
-                if (methodNames.contains(f.name) 
-                    && f.hasCompliantParameters(bindingContext)
-                    && f.listStatements().none { it.throwsException(bindingContext) }) {
-                    kotlinFileContext.reportIssue(f.nameIdentifier ?: f,
-                        "Enable server certificate validation on this SSL/TLS connection.")
-                }
+        if (function.belongsToTrustManagerClass(bindingContext)
+            && !function.callsCheckTrusted(bindingContext)
+            && !function.throwsCertificateExceptionWithoutCatching(bindingContext)
+        ) {
+            kotlinFileContext.reportIssue(function.nameIdentifier ?: function,
+                "Enable server certificate validation on this SSL/TLS connection.")
+        }
+    }
+
+    private fun KtNamedFunction.belongsToTrustManagerClass(bindingContext: BindingContext): Boolean {
+        return funMatchers.any { it.matches(this, bindingContext) }
+    }
+
+    /*
+     * Returns true if a function contains a call to "checkClientTrusted" or "checkServerTrusted".
+     */
+    private fun KtNamedFunction.callsCheckTrusted(bindingContext: BindingContext): Boolean {
+        val visitor = object : KtVisitorVoid() {
+            private var foundCheckTrustedCall: Boolean = false
+
+            override fun visitCallExpression(expression: KtCallExpression) {
+                foundCheckTrustedCall = funMatchers.any { it.matches(expression, bindingContext) }
             }
+
+            fun callsCheckTrusted(): Boolean {
+                return foundCheckTrustedCall
+            }
+        }
+        this.acceptRecursively(visitor)
+        return visitor.callsCheckTrusted()
+    }
+
+    /*
+     * Returns true only when the function throws a CertificateException without a catch against it.
+     */
+    private fun KtNamedFunction.throwsCertificateExceptionWithoutCatching(bindingContext: BindingContext): Boolean {
+        val visitor = ThrowCatchVisitor(bindingContext)
+        this.acceptRecursively(visitor)
+        return visitor.throwsCertificateExceptionWithoutCatching()
+    }
+
+    private class ThrowCatchVisitor(private val bindingContext: BindingContext) : KtVisitorVoid() {
+        private var throwFound: Boolean = false
+        private var catchFound: Boolean = false
+
+        override fun visitThrowExpression(expression: KtThrowExpression) {
+            val callExpr = expression.thrownExpression as? KtCallExpression
+            throwFound = callExpr?.let {
+                ConstructorMatcher {
+                    qualifier = "java.security.cert.CertificateException"
+                    withNoArguments()
+                }.matches(it, bindingContext)
```

Comment:
We should be able to make our lives a lot easier with a couple of little helpers, now and in future, and get more accurate results at the same time. Copy & paste these two helpers into `ApiExtensions.kt`:
```
fun PsiElement?.getType(bindingContext: BindingContext): KotlinType? =
    this?.let {
        when(this) {
            is Call -> getResolvedCall(bindingContext)?.getReturnType()
            is KtCallExpression -> getResolvedCall(bindingContext)?.getReturnType()
            is KtParameter -> determineType(bindingContext)
            is KtTypeReference -> determineType(bindingContext)
            is KtProperty -> determineType(bindingContext)
            is KtDotQualifiedExpression -> resolveReferenceTarget(bindingContext).getType()
            is KtReferenceExpression -> bindingContext.get(BindingContext.REFERENCE_TARGET, this).getType()
            else -> bindingContext.get(BindingContext.VARIABLE, it)?.type
        }

    }

fun DeclarationDescriptor?.getType(): KotlinType? =
    this?.let {
        when(this) {
            is FunctionDescriptor -> returnType
            is ClassDescriptor -> classValueType
            else -> null
        }
    }
```
Then, you can simply replace the body of this function with:
```
            if (
                expression.thrownExpression.getType(bindingContext)
                    ?.getJetTypeFqName(false) == "java.security.cert.CertificateException"
            ) {
                throwFound = true
            }
```

Reply:
These snippets are and will be very useful. Thanks for sharing.